scapy/layers/ipsec.py

```diff
@@ -165,6 +165,270 @@ def data_for_encryption(self):
 
 
+###############################################################################
+if conf.crypto_valid_advanced:
+    from cryptography.exceptions import InvalidTag
+    from cryptography.hazmat.backends import default_backend
+    from cryptography.hazmat.primitives.ciphers import (
+        aead,
+    )
+else:
+    log_loading.info("Can't import python-cryptography v1.7+. "
+                     "Disabled IPsec encryption/authentication.")
+    default_backend = None
+    InvalidTag = Exception
+    aead = None
+
+###############################################################################
+
+
+class AEAD_Algo(object):
+    """
+    IPsec AEAD algorithm
+    """
+
+    def __init__(self, name, algo, block_size=None, iv_size=None,
+                 key_size=None, icv_size=None, salt_size=None):  # noqa: E501
+        """
+        :param name: the name of this encryption algorithm
+        :param algo: a Cipher module
+        :param block_size: the length a block for this algo. Defaults to the
+                           `block_size` of the cipher.
+        :param iv_size: the length of the initialization vector of this algo.
+                        Defaults to the `block_size` of the cipher.
+        :param key_size: an integer or list/tuple of integers. If specified,
+                         force the secret keys length to one of the values.
+                         Defaults to the `key_size` of the cipher.
+        :param icv_size: the length of the Integrity Check Value of this algo.
+                         Used by Combined Mode Algorithms e.g. GCM
+        :param salt_size: the length of the salt to use as the IV prefix.
+                          Usually used by Counter modes e.g. CTR
+        :param format_mode_iv: function to format the Initialization Vector
+                               e.g. handle the salt value
+                               Default is the random buffer from `generate_iv`
+        """
+        self.name = name
+        self.algo = algo
+        self.icv_size = icv_size
+
+        if block_size is not None:
+            self.block_size = block_size
+        else:
+            self.block_size = 1
+
+        if iv_size is None:
+            self.iv_size = self.block_size
+        else:
+            self.iv_size = iv_size
+
+        if key_size is not None:
+            self.key_size = key_size
+        else:
+            self.key_size = None
+
+        if salt_size is None:
+            self.salt_size = 0
+        else:
+            self.salt_size = salt_size
+
+        self._format_mode_iv = lambda sa, iv, **kw: sa.aead_salt + iv
+
+    def check_key(self, key):
+        """
+        Check that the key length is valid.
+
+        :param key:    a byte string
+        """
+        if self.key_size:
+            if not (len(key) == self.key_size or len(key) in self.key_size):
+                raise TypeError('invalid key size %s, must be %s' %
+                                (len(key), self.key_size))
```

Comment:
Could you make `AEAD_Algo` import `AuthAlgo` in order to avoid having to copy/paste such functions?

Reply:
It depends of the bellow question...

Reply:
Hello,

Regarding the way IPSec works, it makes no sense. But I am open to suggestion to avoid duplicated functions (padding, key size check and so on...).